Laganellac/2021_11_1_EnhanceBlueprintExport

Laganellac opened the pull request:
### Type of change

* [ ] Bug fix
* [x] New feature
* [ ] Documentation update
* [ ] Other <!-- please explain with a note below -->

### How Has This Been Tested?

This has been tested on Ubuntu 18.04 using the GUI & CLI to exercise the new options added to the blueprint export. A new blueprint_export.py has been added to the repo with baselines (generated on my workstation) checked in under test/baseline/databases/blueprint_export. Generally, the multi domain silo datasets have been used to test the blueprint flatten / partition functionality; both partitioning them into a greater number of domains or combining them down to a smaller number of domains.

### Checklist:
- [x] I have followed the [style guidelines][1] of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code where applicable.
- [x] I have updated the release notes.
- ~~[ ]  I have made corresponding changes to the documentation.~~
- [x] I have added debugging support to my changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have confirmed new and existing unit tests pass locally with my changes.
- [x] I have added new baselines for any new tests to the repo.
- ~~[ ] I have NOT made any changes to [*protocol* or *public interfaces*][3] in an RC branch.~~
- [ ] I have assigned reviewers (see [VisIt's PR procedures][2] for more information).

[1]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/StyleGuide.html
[2]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/pr_create.html#reviewers
[3]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/RCDevelopment.html#communication-protocols-and-public-apis

### What does this pull request include?

This pull request includes enhancements to the blueprint export plugin which allow users to apply blueprint's "flatten" and "partition" operations to the exported dataset. This pull request includes:

* Updates to blueprint plugin XML to support write options
* The addition of an "operation" option (enum) to the blueprint export which supports the values "Partition", "Flatten_CSV", and "Flatten_HDF5"
* The addition of a "Partition target number of domains" option which tells blueprint::mesh::partition how many domains to create
* Enhancements to avtBlueprintWriter.C to handle these options. When using partition or flatten, the operations are deferred until CloseFile so that the parallel implementations of flatten / partition can be used safely.
* Fixed a bug in "HomogeneousShapeTopologyToVTKCellArray" which was causing it to needlessly copy the connectivity array repeatedly
* Added avtBlueprintDataAdapter::VTKFieldNameToBlueprint which filters out '/'s from field names before adding them to blueprint. I noticed that invalid blueprint "fields" entries were being created due to field names like "topology/field0" being passed to conduit.
* Added blueprint_export.py which tests partitioning the multi domain silo datasets against image baselines, and flattening the multi domain silo datasets against text (CSV) baselines.
* blueprint_export.py is currently on the skip list because the visit-thirdparty Conduit version will need to be updated to support these features
* Enhancements to the CMake logic which check for partition/flatten support (checks if conduit version is >= 0.8.0). Conditional compilation added to blueprint plugin which disables the new features if conduit version is too old
* Update bv_conduit for version v0.8.0, also have it write the conduit version to the host-config.cmake file so that the above check can be made

Please let me know if you have any questions / need me to make any changes.

cdl-ilight committed:
Add support for flattening a blueprint mesh on write

cdl-ilight committed:
Add support for partitioning a mesh

cdl-ilight committed:
Filter '/' out of field names when given to blueprint

cdl-ilight committed:
No longer copy entire array on every iteration

cdl-ilight committed:
Ghost zones don't participate in partition

cdl-ilight committed:
Add test cases for partition and flatten

cdl-ilight committed:
Add flatten tests

cdl-ilight committed:
Integrate test with visit test suite

cdl-ilight committed:
Add blueprint_export baselines for serial and parallel

cdl-ilight committed:
Merge branch 'develop' into laganellac/2021_11_1_EnhanceBlueprintExport

cdl-ilight committed:
Add to the 3.3 release notes

cdl-ilight committed:
Add conditional compilation for new blueprint features

cdl-ilight committed:
Skip the new test until conduit version is updated in bv

cdl-ilight committed:
Make sure build visit writes the conduit version into host file

cdl-ilight committed:
Update conduit version in bv_conduit

Includes updated checksums

cdl-ilight committed:
Merge branch 'develop' into laganellac/2021_11_1_EnhanceBlueprintExport

Laganellac commented:
Tagging @cyrush and @BradWhitlock as this goes along with some work we've been doing in the Conduit repo.

cyrush commented on src/databases/Blueprint/avtBlueprintDataAdaptor.C:
```diff
@@ -600,22 +614,23 @@ HomogeneousShapeTopologyToVTKCellArray(const Node &n_topo,
-        for (int i = 0 ; i < ncells; i++)
+        Node n_tmp;
+        if(n_topo["elements/connectivity"].dtype().is_int())
         {
```
thank you for fix this!!!!

cyrush replied:
FYI -- I am going to get this on our RC branch for our imminent release 3.2.2

cdl-ilight committed:
Remove the 'v' from the version string printed to cmake

cdl-ilight committed:
Conditionally compile BuildSelections

The function is only used by the partition operation and uses some of
the newer functionality provided by conduit v0.8.0

cyrush reviewed (approved):
Looks good, thanks for wiring all this in!
We will use new conduit for VisIt 3.3 and then folks will have access to these features.

Laganellac commented:
Hey @cyrush, I've recently done some work that builds off the work done in this branch. Specifically, I've added the ability for the user to provide a "MultiLineString" through the DBOptions attributes which is then parsed into a Conduit node in the BlueprintWriter. This node is then passed as the options input to either Flatten or Partition.

The work exists here: https://github.com/Laganellac/visit/tree/laganellac/2022_02_08_AddMultiLineEdit

I was wondering how you would like to organize this. I notice that this PR has already been approved but is not yet merged. I could wait until this gets merged into develop and then create a new PR with the new work OR I could merge my new branch into this existing PR (which would need to be reviewed again, I assume).

Let me know what you think.

cyrush committed:
Merge branch 'develop' into laganellac/2021_11_1_EnhanceBlueprintExport

cyrush commented:
@Laganellac  This looks good, thanks! I plan to merge this today, then you can do a PR to develop for the extended work.

cyrush committed:
Merge branch 'develop' into laganellac/2021_11_1_EnhanceBlueprintExport

cyrush commented:
@Laganellac merged, so you can move forward with the other PR into `develop`.
